### ISSUE TITLE
Eloquent/Model::withCount doesn't specify database/schema name

### ISSUE THREAD

**Issue by foreshadow** [laravel/framework#23042]
- Laravel Version: 5.5.33
- PHP Version: 7.1.8
- Database Driver & Version: Mysql 5.7

#### Eloquent/Model::withCount doesn't specify database/schema name
I have 2 models User and Problem. Their table locates in different database/schema in same server (`users` locates in connection `mysql` a.k.a. schema `acm`, `problem` locates in connection `mysql2` a.k.a. schema `boj` in my case).
I have defined `protected $connection` for these two models. Anything other works well. 
And when I call `User::with('solved')` it works well, but when calling `User::withCount('solved')`, the generated sql sentence doesn't specify the `problem` table's database/schema, which will cause an error.

(formatted)
```
(3/3) QueryException
SQLSTATE[42S02]: Base table or view not found: 1146 Table 'acm.problem' doesn't exist (SQL: 
```
```sql
select `users`.*, (
    select count(*) from `problem` 
    inner join `boj`.`solution` on `problem`.`problem_id` = `boj`.`solution`.`problem_id` 
    where `users`.`id` = `boj`.`solution`.`acm_user_id` and `boj`.`solution`.`result` = 4
) as `solved_count` 
from `users` order by `active_at` desc limit 150 offset 0)
```

#### Defines
```php
class User
{
    protected $connection = 'mysql';
    // some other defines 

    public function solved()
    {
        $mysql2 = config('database.connections.mysql2.database');
        return $this
            ->belongsToMany(Problem::class, "$mysql2.solution", 'acm_user_id', 'problem_id')
            ->wherePivot('result', 4);
    }
}
```
I believe it doen't matter where the `solution` table locates and how I call it... Although it may looks ugly.
```php
class Problem
{
    protected $connection = 'mysql2';
    protected $table = 'problem';
    // some other defines
}
```
#### Call
```php
User::with('roles')->withCount('solved')->orderBy('active_at', 'desc')->paginate(150);
```

**Comment by foreshadow** [comment]
Besides, `withCount` a `hasMany` relation will cause the same problem.

**Comment by miki131** [comment]
@foreshadow, try this
```php
User::with('roles')
  ->withCount(['solved' => function($query) {
    $mysql2 = config('database.connections.mysql2.database');
    $query->from("$mysql2.problem");
  }])
  ->orderBy('active_at', 'desc')
  ->paginate(150);
```

**Comment by vesper8** [comment]
@miki131's approach works for me. However.. the odd thing is that if you use with() instead of withCount.. on top of with() not requiring to specify the 2nd db connection.. running ->count() on the related model fetched from with() is at least 2x as fast as when using withCount

withCount is supposed to be faster, but it seems that doing it the way @miki131 proposed, while it works, it seems less optimized somehow and is much slower

**Comment by driesvints** [comment]
@nowox we're open to PRs for this. I'm working my way through all of the bugs but it'll take time.

**Comment by staudenmeir** [comment]
In the meantime, you can use this package: https://github.com/hoyvoy/laravel-cross-database-subqueries

**Comment by alwintom** [comment]
I have found a quick fix. I modified `Illuminate\Database\Eloquent\Builder::setModel` method like this:

````php
    /**
     * Set a model instance for the model being queried.
     *
     * @param  \Illuminate\Database\Eloquent\Model  $model
     * @return $this
     */
    public function setModel(Model $model)
    {
        $this->model = $model;

        $this->query->from($model->getConnection()->getDatabaseName() . '.' . $model->getTable());

        return $this;
    }
````

What I did was force prepend the name of the database to the table name within `FROM`. The db name is obtained using the model's connection name.

I made this change to fix #29125 

I applied this change and so far my application is working fine and it does not seem to have broken anything that was working. The only issue I see is the rather longer queries.

I could create a PR if this fix does not break anything

**Comment by driesvints** [comment]
@alwintom the pr you sent in seems too much of a breaking change. What do you think @staudenmeir ?

**Comment by alwintom** [comment]
I think so too. I have alreday closed the PR after realizing that. During
testing against sqlite DBs it failed.

However i am going to keep it for my application as we strictly use mysql
and it seems to work for us

**Comment by skalero01** [comment]
> ```
> $this->query->from($model->getConnection()->getDatabaseName() . '.' . $model->getTable());
> ```

It works, if it fails with sqlite why dont add a validation only to check that is not sqlite? Or what other problems happens with using this code?

**Comment by driesvints** [comment]
I've been trying to look into this again but haven't made much progress so far. I *think* the problem is that the `newQuery` returned from below doesn't have the custom connection from the related model defined but I cannot figure out why that is.

https://github.com/laravel/framework/blob/7.x/src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php#L386

Appreciating any help with figuring this out.

**Comment by ryangjchandler** [comment]
> I've been trying to look into this again but haven't made much progress so far. I _think_ the problem is that the `newQuery` returned from below doesn't have the custom connection from the related model defined but I cannot figure out why that is.
> 
> https://github.com/laravel/framework/blob/7.x/src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php#L386
> 
> Appreciating any help with figuring this out.

Doesn't this eventually just call `getTable` on the `Model`, it never actually resolves the connection? https://github.com/laravel/framework/blob/4e9f584689a185b2d10a543bf20fbca5e4e2989c/src/Illuminate/Database/Eloquent/Builder.php#L1270

**Comment by ryangjchandler** [comment]
One workaround for this at the moment is prefixing the table with the database name:

```
class Problem extends Model
{
    protected $table = 'mysql2.problems';
} 
```

**Comment by alwintom** [comment]
> 
> 
> One workaround for this at the moment is prefixing the table with the database name:
> 
> ```
> class Problem extends Model
> {
>     protected $table = 'mysql2.problem';
> } 
> ```

This will result in the builder generating queries like this:
`SELECT ..... from ::memory::.users`
when you use an in-memory sqlite database.

That being said, I resorted to a solution similar to yours in my application and added a ENV check to make sure my in-memory sqlite based unit tests to pass as well.

**Comment by yaim** [comment]
@driesvints I followed your lead but i guess that part is OK. I think the problem is that the [`selectSub`](https://github.com/laravel/framework/blob/7.x/src/Illuminate/Database/Eloquent/Concerns/QueriesRelationships.php#L408) is transferring the query object into raw SQL. That drops the custom connection from the relation. I couldn't figure out how to fix it. :man_shrugging:

**Comment by staudenmeir** [comment]
In principle, the package's implementation is the right approach:
https://github.com/hoyvoy/laravel-cross-database-subqueries/blob/master/src/Eloquent/Concerns/QueriesRelationships.php#L84

We compare the connections and prefix the subquery's table name if they are different. I only find the temporary separator `<-->` to be a weird choice, not sure why they implemented it this way.

The trait also uses the same approach to fix the issue for `has()`/`whereHas()` subqueries. It makes sense to fix both issues in one go and extract the necessary code into a shared helper method.

We also need to consider the package's open [issues](https://github.com/hoyvoy/laravel-cross-database-subqueries/issues), especially [this](https://github.com/hoyvoy/laravel-cross-database-subqueries/issues/12) one.

**Comment by yaim** [comment]
@staudenmeir but this won't support `:memory:`, right?

**Comment by staudenmeir** [comment]
@yaim No. IMO, cross-database queries on SQLite are _way_ too much of an edge case.

They are [possible](https://www.sqlite.org/lang_attach.html) for file-based databases and apparently even for in-memory ones, but I don't think that would work with Laravel.

**Comment by MwSpaceLLC** [comment]
Hi there, we want to write that for us all work fine in mysql with package [cross-database](https://github.com/hoyvoy/laravel-cross-database-subqueries).

![image](https://user-images.githubusercontent.com/29952045/86354491-70e70e00-bc69-11ea-9ef7-e801cdfb899b.png)

In this query User are in DB1 and categories are in DB2. the result work great!

![image](https://user-images.githubusercontent.com/29952045/86354653-b572a980-bc69-11ea-9b87-4858c756ac98.png)

For us, we must specify relation connection ('database.relation') in : 

![image](https://user-images.githubusercontent.com/29952045/86354747-df2bd080-bc69-11ea-8420-17976b69dc74.png)

And for User, we must replace Authenticatable with new model class:

![image](https://user-images.githubusercontent.com/29952045/86354822-01255300-bc6a-11ea-928f-60c026b7ec38.png)

![image](https://user-images.githubusercontent.com/29952045/86354850-0bdfe800-bc6a-11ea-9f53-90bbf9862258.png)

Thanks for the trick :+1:

**Comment by ollieread** [comment]
I may be totally off the mark, but I believe is down to Laravel overriding the connection of the related model.

Let's take the `hasMany` relation as an example. If I have `Model1` using connection `mysql1` that has many `Model2` using `mysql2`, the `Model::$connection` property will be overridden once I attempt to get an instance of the `Relation` class.

If `Model1` has the following method.

```php
public function model2() {
    return $this->hasMany(Model2::class);
}
```

When I call `$model1->model2()` the first thing that `HasRelationships::hasMany()` calls is `HasRelationships::newRelatedInstance()`. In Laravel 7, the `newRelatedInstance()` method has the following code:

```php
protected function newRelatedInstance($class)
{
    return tap(new $class, function ($instance) {
        if (! $instance->getConnectionName()) {
            $instance->setConnection($this->connection);
        }
    });
}
```

Since `HasRelationships` is on a model1 the following happens:

- `Model1::model2()` calls...
- `Model1::hasMany()`, which calls...
- `Model1::newRelatedInstance()` which...
- Creates a new instance of `Model2`, then
- Calls `Model2::setConnection()` with the value of `Model1::$connection`.

I've ran into this particular thing a few times. There's an unrelated but similar thing with `setTable()` usage.

I'm fairly confident this is the source of the issue @driesvints, because the second model isn't aware of a new connection, as it was overridden.

**Comment by ollieread** [comment]
I've done some testing on this, and I can't recreate the issue you're having @driesvints. The query created in `QueriesRelationships::withCount()` definitely has the new connection. The problem is that the query is turned into its raw SQL and added as a subquery.

Looking at it, the best solution I can see is one where the database name is appended to the from in [`Builder::parseSub()`](https://github.com/laravel/framework/blob/7.x/src/Illuminate/Database/Query/Builder.php#L339).

**Comment by driesvints** [comment]
@foreshadow @MwSpaceLLC does the solution from @ollieread work for you?

**Comment by MwSpaceLLC** [comment]
@driesvints for now we use package cross-database.😁

**Comment by driesvints** [comment]
@MwSpaceLLC yes but I'm asking that if your problem works if you apply the patch from @ollieread. Without that package installed.

**Comment by MwSpaceLLC** [comment]
@driesvints now the project is under production and we not have any time for testing. when i have some time to test i reply here

**Comment by driesvints** [comment]
To all on this thread: I'm going to leave this open for a few days to give any of you the time to test @ollieread's solution. After that I'll close it. We're of course still accepting PRs if anyone's up for it. Thanks

**Comment by ollieread** [comment]
@driesvints I've tested it locally and all seems to work for me. I'm happy to submit a PR with that solution.

**Comment by driesvints** [comment]
@ollieread could do that yeah. But we'll need to make sure we don't break any tests.

**Comment by driesvints** [comment]
Send to 6.x please, thanks!

**Comment by ourfamilygithub** [comment]
> I've done some testing on this, and I can't recreate the issue you're having @driesvints. The query created in `QueriesRelationships::withCount()` definitely has the new connection. The problem is that the query is turned into its raw SQL and added as a subquery.
> 
> Looking at it, the best solution I can see is one where the database name is appended to the from in [`Builder::parseSub()`](https://github.com/laravel/framework/blob/7.x/src/Illuminate/Database/Query/Builder.php#L339). The method could be made to be something like this.
> 
> ```
> protected function parseSub($query)
> {
>     if ($query instanceof self || $query instanceof EloquentBuilder || $query instanceof Relation) {
>         if ($query->getConnection()->getDatabaseName() !== $this->getConnection()->getDatabaseName()) {
>             $databaseName = $query->getConnection()->getDatabaseName();
> 
> 			if (strpos($query->from, $databaseName) !== 0) {
> 				$query->from($databaseName .'.'.$query->from);
>             }
>         }
> 
>         return [$query->toSql(), $query->getBindings()];
>     } elseif (is_string($query)) {
>         return [$query, []];
>     } else {
>         throw new InvalidArgumentException(
>             'A subquery must be a query builder instance, a Closure, or a string.'
>         );
>     }
> }
> ```
> 
> I can't think of anything prettier. This should have the benefit that it'll work for any cross-database subqueries, as long as they're not on totally different connections. It could perhaps check for the presence of `.` instead of the database name.
> 
> I can make a PR if you wish.

@ollieread , @driesvints , Thank you this works very well in my case

**Comment by ollieread** [comment]
@driesvints The fix for this has now been merged in (#33755) so this should now be sorted.

**Comment by driesvints** [comment]
Thanks @ollieread!
